File: src/csharp/Intel/Iced.UnitTests/Intel/EncoderTests/OpCodeInfoTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && OPCODE_INFO
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.EncoderTests {
	public sealed class OpCodeInfoTests {
		[Theory]
		[MemberData(nameof(TestAllOpCodeInfos_Data))]
#pragma warning disable xUnit1026 // Theory methods should use all of their parameters
		void Test_all_OpCodeInfos(int lineNo, Code code, string opCodeString, string instructionString, OpCodeInfoTestCase tc) {
#pragma warning restore xUnit1026 // Theory methods should use all of their parameters
			var info = tc.Code.ToOpCode();
			Assert.Equal(tc.Code, info.Code);
#pragma warning disable xUnit2006 // Do not use invalid string equality check
			// Show the full string without ellipses by using Equal<string>() instead of Equal()
			Assert.Equal<string>(tc.OpCodeString, info.ToOpCodeString());
			Assert.Equal<string>(tc.InstructionString, info.ToInstructionString());
#pragma warning restore xUnit2006 // Do not use invalid string equality check
			Assert.True((object)info.ToInstructionString() == info.ToString());
			Assert.Equal(tc.Mnemonic, info.Mnemonic);
			Assert.Equal(tc.Encoding, info.Encoding);
			Assert.Equal(tc.IsInstruction, info.IsInstruction);
			Assert.Equal(tc.Mode16, info.Mode16);
			Assert.Equal(tc.Mode16, info.IsAvailableInMode(16));
			Assert.Equal(tc.Mode32, info.Mode32);
			Assert.Equal(tc.Mode32, info.IsAvailableInMode(32));
			Assert.Equal(tc.Mode64, info.Mode64);
			Assert.Equal(tc.Mode64, info.IsAvailableInMode(64));
			Assert.Equal(tc.Fwait, info.Fwait);
			Assert.Equal(tc.OperandSize, info.OperandSize);
			Assert.Equal(tc.AddressSize, info.AddressSize);
			Assert.Equal(tc.L, info.L);
			Assert.Equal(tc.W, info.W);
			Assert.Equal(tc.IsLIG, info.IsLIG);
			Assert.Equal(tc.IsWIG, info.IsWIG);
			Assert.Equal(tc.IsWIG32, info.IsWIG32);
			Assert.Equal(tc.TupleType, info.TupleType);
			Assert.Equal(tc.MemorySize, info.MemorySize);
			Assert.Equal(tc.BroadcastMemorySize, info.BroadcastMemorySize);
			Assert.Equal(tc.DecoderOption, info.DecoderOption);
			Assert.Equal(tc.CanBroadcast, info.CanBroadcast);
			Assert.Equal(tc.CanUseRoundingControl, info.CanUseRoundingControl);
			Assert.Equal(tc.CanSuppressAllExceptions, info.CanSuppressAllExceptions);
			Assert.Equal(tc.CanUseOpMaskRegister, info.CanUseOpMaskRegister);
			Assert.Equal(tc.RequireOpMaskRegister, info.RequireOpMaskRegister);
			if (tc.RequireOpMaskRegister) {
				Assert.True(info.CanUseOpMaskRegister);
				Assert.False(info.CanUseZeroingMasking);
			}
			Assert.Equal(tc.CanUseZeroingMasking, info.CanUseZeroingMasking);
			Assert.Equal(tc.CanUseLockPrefix, info.CanUseLockPrefix);
			Assert.Equal(tc.CanUseXacquirePrefix, info.CanUseXacquirePrefix);
			Assert.Equal(tc.CanUseXreleasePrefix, info.CanUseXreleasePrefix);
			Assert.Equal(tc.CanUseRepPrefix, info.CanUseRepPrefix);
			Assert.Equal(tc.CanUseRepnePrefix, info.CanUseRepnePrefix);
			Assert.Equal(tc.CanUseBndPrefix, info.CanUseBndPrefix);
			Assert.Equal(tc.CanUseHintTakenPrefix, info.CanUseHintTakenPrefix);
			Assert.Equal(tc.CanUseNotrackPrefix, info.CanUseNotrackPrefix);
			Assert.Equal(tc.IgnoresRoundingControl, info.IgnoresRoundingControl);
			Assert.Equal(tc.AmdLockRegBit, info.AmdLockRegBit);
			Assert.Equal(tc.DefaultOpSize64, info.DefaultOpSize64);
			Assert.Equal(tc.ForceOpSize64, info.ForceOpSize64);
			Assert.Equal(tc.IntelForceOpSize64, info.IntelForceOpSize64);
			Assert.Equal(tc.Cpl0 && !tc.Cpl1 && !tc.Cpl2 && !tc.Cpl3, info.MustBeCpl0);
			Assert.Equal(tc.Cpl0, info.Cpl0);
			Assert.Equal(tc.Cpl1, info.Cpl1);
			Assert.Equal(tc.Cpl2, info.Cpl2);
			Assert.Equal(tc.Cpl3, info.Cpl3);
			Assert.Equal(tc.IsInputOutput, info.IsInputOutput);
			Assert.Equal(tc.IsNop, info.IsNop);
			Assert.Equal(tc.IsReservedNop, info.IsReservedNop);
			Assert.Equal(tc.IsSerializingIntel, info.IsSerializingIntel);
			Assert.Equal(tc.IsSerializingAmd, info.IsSerializingAmd);
			Assert.Equal(tc.MayRequireCpl0, info.MayRequireCpl0);
			Assert.Equal(tc.IsCetTracked, info.IsCetTracked);
			Assert.Equal(tc.IsNonTemporal, info.IsNonTemporal);
			Assert.Equal(tc.IsFpuNoWait, info.IsFpuNoWait);
			Assert.Equal(tc.IgnoresModBits, info.IgnoresModBits);
			Assert.Equal(tc.No66, info.No66);
			Assert.Equal(tc.NFx, info.NFx);
			Assert.Equal(tc.RequiresUniqueRegNums, info.RequiresUniqueRegNums);
			Assert.Equal(tc.RequiresUniqueDestRegNum, info.RequiresUniqueDestRegNum);
			Assert.Equal(tc.IsPrivileged, info.IsPrivileged);
			Assert.Equal(tc.IsSaveRestore, info.IsSaveRestore);
			Assert.Equal(tc.IsStackInstruction, info.IsStackInstruction);
			Assert.Equal(tc.IgnoresSegment, info.IgnoresSegment);
			Assert.Equal(tc.IsOpMaskReadWrite, info.IsOpMaskReadWrite);
			Assert.Equal(tc.RealMode, info.RealMode);
			Assert.Equal(tc.ProtectedMode, info.ProtectedMode);
			Assert.Equal(tc.Virtual8086Mode, info.Virtual8086Mode);
			Assert.Equal(tc.CompatibilityMode, info.CompatibilityMode);
			Assert.Equal(tc.LongMode, info.LongMode);
			Assert.Equal(tc.UseOutsideSmm, info.UseOutsideSmm);
			Assert.Equal(tc.UseInSmm, info.UseInSmm);
			Assert.Equal(tc.UseOutsideEnclaveSgx, info.UseOutsideEnclaveSgx);
			Assert.Equal(tc.UseInEnclaveSgx1, info.UseInEnclaveSgx1);
			Assert.Equal(tc.UseInEnclaveSgx2, info.UseInEnclaveSgx2);
			Assert.Equal(tc.UseOutsideVmxOp, info.UseOutsideVmxOp);
			Assert.Equal(tc.UseInVmxRootOp, info.UseInVmxRootOp);
			Assert.Equal(tc.UseInVmxNonRootOp, info.UseInVmxNonRootOp);
			Assert.Equal(tc.UseOutsideSeam, info.UseOutsideSeam);
			Assert.Equal(tc.UseInSeam, info.UseInSeam);
			Assert.Equal(tc.TdxNonRootGenUd, info.TdxNonRootGenUd);
			Assert.Equal(tc.TdxNonRootGenVe, info.TdxNonRootGenVe);
			Assert.Equal(tc.TdxNonRootMayGenEx, info.TdxNonRootMayGenEx);
			Assert.Equal(tc.IntelVmExit, info.IntelVmExit);
			Assert.Equal(tc.IntelMayVmExit, info.IntelMayVmExit);
			Assert.Equal(tc.IntelSmmVmExit, info.IntelSmmVmExit);
			Assert.Equal(tc.AmdVmExit, info.AmdVmExit);
			Assert.Equal(tc.AmdMayVmExit, info.AmdMayVmExit);
			Assert.Equal(tc.TsxAbort, info.TsxAbort);
			Assert.Equal(tc.TsxImplAbort, info.TsxImplAbort);
			Assert.Equal(tc.TsxMayAbort, info.TsxMayAbort);
			Assert.Equal(tc.IntelDecoder16, info.IntelDecoder16);
			Assert.Equal(tc.IntelDecoder32, info.IntelDecoder32);
			Assert.Equal(tc.IntelDecoder64, info.IntelDecoder64);
			Assert.Equal(tc.AmdDecoder16, info.AmdDecoder16);
			Assert.Equal(tc.AmdDecoder32, info.AmdDecoder32);
			Assert.Equal(tc.AmdDecoder64, info.AmdDecoder64);
			Assert.Equal(tc.Table, info.Table);
			Assert.Equal(tc.MandatoryPrefix, info.MandatoryPrefix);
			Assert.Equal(tc.OpCode, info.OpCode);
			Assert.Equal(tc.OpCodeLength, info.OpCodeLength);
			Assert.Equal(tc.IsGroup, info.IsGroup);
			Assert.Equal(tc.GroupIndex, info.GroupIndex);
			Assert.Equal(tc.IsRmGroup, info.IsRmGroup);
			Assert.Equal(tc.RmGroupIndex, info.RmGroupIndex);
			Assert.Equal(tc.OpCount, info.OpCount);
			Assert.Equal(tc.Op0Kind, info.Op0Kind);
			Assert.Equal(tc.Op1Kind, info.Op1Kind);
			Assert.Equal(tc.Op2Kind, info.Op2Kind);
			Assert.Equal(tc.Op3Kind, info.Op3Kind);
			Assert.Equal(tc.Op4Kind, info.Op4Kind);
			Assert.Equal(tc.Op0Kind, info.GetOpKind(0));
			Assert.Equal(tc.Op1Kind, info.GetOpKind(1));
			Assert.Equal(tc.Op2Kind, info.GetOpKind(2));
			Assert.Equal(tc.Op3Kind, info.GetOpKind(3));
			Assert.Equal(tc.Op4Kind, info.GetOpKind(4));
			Static.Assert(IcedConstants.MaxOpCount == 5 ? 0 : -1);
			for (int i = tc.OpCount; i < IcedConstants.MaxOpCount; i++)
				Assert.Equal(OpCodeOperandKind.None, info.GetOpKind(i));
#if MVEX
			Assert.Equal(tc.Mvex.EHBit, info.MvexEHBit);
			Assert.Equal(tc.Mvex.CanUseEvictionHint, info.MvexCanUseEvictionHint);
			Assert.Equal(tc.Mvex.CanUseImmRoundingControl, info.MvexCanUseImmRoundingControl);
			Assert.Equal(tc.Mvex.IgnoresOpMaskRegister, info.MvexIgnoresOpMaskRegister);
			Assert.Equal(tc.Mvex.NoSaeRc, info.MvexNoSaeRc);
			Assert.Equal(tc.Mvex.TupleTypeLutKind, info.MvexTupleTypeLutKind);
			Assert.Equal(tc.Mvex.ConversionFunc, info.MvexConversionFunc);
			Assert.Equal(tc.Mvex.ValidConversionFuncsMask, info.MvexValidConversionFuncsMask);
			Assert.Equal(tc.Mvex.ValidSwizzleFuncsMask, info.MvexValidSwizzleFuncsMask);
#endif
		}
		public static IEnumerable<object[]> TestAllOpCodeInfos_Data => OpCodeInfoTestCases.OpCodeInfoTests.Select(a => new object[] { a.LineNumber, a.Code, a.OpCodeString, a.InstructionString, a });

		[Fact]
		void GetOpKindThrowsIfInvalidInput() {
			var info = Code.Aaa.ToOpCode();
			Assert.Throws<ArgumentOutOfRangeException>(() => info.GetOpKind(int.MinValue));
			Assert.Throws<ArgumentOutOfRangeException>(() => info.GetOpKind(-1));
			info.GetOpKind(0);
			info.GetOpKind(IcedConstants.MaxOpCount - 1);
			Assert.Throws<ArgumentOutOfRangeException>(() => info.GetOpKind(IcedConstants.MaxOpCount));
			Assert.Throws<ArgumentOutOfRangeException>(() => info.GetOpKind(int.MaxValue));
		}

		[Fact]
		void Verify_Instruction_OpCodeInfo() {
			for (int i = 0; i < IcedConstants.CodeEnumCount; i++) {
				var code = (Code)i;
				Instruction instruction = default;
				instruction.Code = (Code)i;
				Assert.True(ReferenceEquals(code.ToOpCode(), instruction.OpCode));
			}
		}

		[Fact]
		void Make_sure_all_Code_values_are_tested_exactly_once() {
			var tested = new bool[IcedConstants.CodeEnumCount];
			foreach (var info in OpCodeInfoTestCases.OpCodeInfoTests) {
				Assert.False(tested[(int)info.Code]);
				tested[(int)info.Code] = true;
			}
			var sb = new StringBuilder();
			var codeNames = ToEnumConverter.GetCodeNames();
			for (int i = 0; i < tested.Length; i++) {
				if (!tested[i] && !CodeUtils.IsIgnored(codeNames[i])) {
					if (sb.Length > 0)
						sb.Append(',');
					sb.Append(codeNames[i]);
				}
			}
			Assert.Equal(string.Empty, sb.ToString());
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/OpCodeInfoTests.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.Code;
import com.github.icedland.iced.x86.CodeUtils;
import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.ToCode;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class OpCodeInfoTests {
	@ParameterizedTest
	@MethodSource("test_all_OpCodeInfos_Data")
	void test_all_OpCodeInfos(int lineNo, int code, String opCodeString, String instructionString, OpCodeInfoTestCase tc) {
		OpCodeInfo info = Code.toOpCode(tc.code);
		assertEquals(tc.code, info.getCode());
		assertEquals(tc.opCodeString, info.toOpCodeString());
		assertEquals(tc.instructionString, info.toInstructionString());
		assertTrue(info.toInstructionString() == info.toString());
		assertEquals(tc.mnemonic, info.getMnemonic());
		assertEquals(tc.encoding, info.getEncoding());
		assertEquals(tc.isInstruction, info.isInstruction());
		assertEquals(tc.mode16, info.getMode16());
		assertEquals(tc.mode16, info.isAvailableInMode(16));
		assertEquals(tc.mode32, info.getMode32());
		assertEquals(tc.mode32, info.isAvailableInMode(32));
		assertEquals(tc.mode64, info.getMode64());
		assertEquals(tc.mode64, info.isAvailableInMode(64));
		assertEquals(tc.fwait, info.getFwait());
		assertEquals(tc.operandSize, info.getOperandSize());
		assertEquals(tc.addressSize, info.getAddressSize());
		assertEquals(tc.l, info.getL());
		assertEquals(tc.w, info.getW());
		assertEquals(tc.isLIG, info.isLIG());
		assertEquals(tc.isWIG, info.isWIG());
		assertEquals(tc.isWIG32, info.isWIG32());
		assertEquals(tc.tupleType, info.getTupleType());
		assertEquals(tc.memorySize, info.getMemorySize());
		assertEquals(tc.broadcastMemorySize, info.getBroadcastMemorySize());
		assertEquals(tc.decoderOption, info.getDecoderOption());
		assertEquals(tc.canBroadcast, info.canBroadcast());
		assertEquals(tc.canUseRoundingControl, info.canUseRoundingControl());
		assertEquals(tc.canSuppressAllExceptions, info.canSuppressAllExceptions());
		assertEquals(tc.canUseOpMaskRegister, info.canUseOpMaskRegister());
		assertEquals(tc.requireOpMaskRegister, info.getRequireOpMaskRegister());
		if (tc.requireOpMaskRegister) {
			assertTrue(info.canUseOpMaskRegister());
			assertFalse(info.canUseZeroingMasking());
		}
		assertEquals(tc.canUseZeroingMasking, info.canUseZeroingMasking());
		assertEquals(tc.canUseLockPrefix, info.canUseLockPrefix());
		assertEquals(tc.canUseXacquirePrefix, info.canUseXacquirePrefix());
		assertEquals(tc.canUseXreleasePrefix, info.canUseXreleasePrefix());
		assertEquals(tc.canUseRepPrefix, info.canUseRepPrefix());
		assertEquals(tc.canUseRepnePrefix, info.canUseRepnePrefix());
		assertEquals(tc.canUseBndPrefix, info.canUseBndPrefix());
		assertEquals(tc.canUseHintTakenPrefix, info.canUseHintTakenPrefix());
		assertEquals(tc.canUseNotrackPrefix, info.canUseNotrackPrefix());
		assertEquals(tc.ignoresRoundingControl, info.getIgnoresRoundingControl());
		assertEquals(tc.amdLockRegBit, info.getAmdLockRegBit());
		assertEquals(tc.defaultOpSize64, info.getDefaultOpSize64());
		assertEquals(tc.forceOpSize64, info.getForceOpSize64());
		assertEquals(tc.intelForceOpSize64, info.getIntelForceOpSize64());
		assertEquals(tc.cpl0 && !tc.cpl1 && !tc.cpl2 && !tc.cpl3, info.getMustBeCpl0());
		assertEquals(tc.cpl0, info.getCpl0());
		assertEquals(tc.cpl1, info.getCpl1());
		assertEquals(tc.cpl2, info.getCpl2());
		assertEquals(tc.cpl3, info.getCpl3());
		assertEquals(tc.isInputOutput, info.isInputOutput());
		assertEquals(tc.isNop, info.isNop());
		assertEquals(tc.isReservedNop, info.isReservedNop());
		assertEquals(tc.isSerializingIntel, info.isSerializingIntel());
		assertEquals(tc.isSerializingAmd, info.isSerializingAmd());
		assertEquals(tc.mayRequireCpl0, info.getMayRequireCpl0());
		assertEquals(tc.isCetTracked, info.isCetTracked());
		assertEquals(tc.isNonTemporal, info.isNonTemporal());
		assertEquals(tc.isFpuNoWait, info.isFpuNoWait());
		assertEquals(tc.ignoresModBits, info.getIgnoresModBits());
		assertEquals(tc.no66, info.getNo66());
		assertEquals(tc.nFx, info.getNFx());
		assertEquals(tc.requiresUniqueRegNums, info.getRequiresUniqueRegNums());
		assertEquals(tc.requiresUniqueDestRegNum, info.getRequiresUniqueDestRegNum());
		assertEquals(tc.isPrivileged, info.isPrivileged());
		assertEquals(tc.isSaveRestore, info.isSaveRestore());
		assertEquals(tc.isStackInstruction, info.isStackInstruction());
		assertEquals(tc.ignoresSegment, info.getIgnoresSegment());
		assertEquals(tc.isOpMaskReadWrite, info.isOpMaskReadWrite());
		assertEquals(tc.realMode, info.getRealMode());
		assertEquals(tc.protectedMode, info.getProtectedMode());
		assertEquals(tc.virtual8086Mode, info.getVirtual8086Mode());
		assertEquals(tc.compatibilityMode, info.getCompatibilityMode());
		assertEquals(tc.longMode, info.getLongMode());
		assertEquals(tc.useOutsideSmm, info.getUseOutsideSmm());
		assertEquals(tc.useInSmm, info.getUseInSmm());
		assertEquals(tc.useOutsideEnclaveSgx, info.getUseOutsideEnclaveSgx());
		assertEquals(tc.useInEnclaveSgx1, info.getUseInEnclaveSgx1());
		assertEquals(tc.useInEnclaveSgx2, info.getUseInEnclaveSgx2());
		assertEquals(tc.useOutsideVmxOp, info.getUseOutsideVmxOp());
		assertEquals(tc.useInVmxRootOp, info.getUseInVmxRootOp());
		assertEquals(tc.useInVmxNonRootOp, info.getUseInVmxNonRootOp());
		assertEquals(tc.useOutsideSeam, info.getUseOutsideSeam());
		assertEquals(tc.useInSeam, info.getUseInSeam());
		assertEquals(tc.tdxNonRootGenUd, info.getTdxNonRootGenUd());
		assertEquals(tc.tdxNonRootGenVe, info.getTdxNonRootGenVe());
		assertEquals(tc.tdxNonRootMayGenEx, info.getTdxNonRootMayGenEx());
		assertEquals(tc.intelVmExit, info.getIntelVmExit());
		assertEquals(tc.intelMayVmExit, info.getIntelMayVmExit());
		assertEquals(tc.intelSmmVmExit, info.getIntelSmmVmExit());
		assertEquals(tc.amdVmExit, info.getAmdVmExit());
		assertEquals(tc.amdMayVmExit, info.getAmdMayVmExit());
		assertEquals(tc.tsxAbort, info.getTsxAbort());
		assertEquals(tc.tsxImplAbort, info.getTsxImplAbort());
		assertEquals(tc.tsxMayAbort, info.getTsxMayAbort());
		assertEquals(tc.intelDecoder16, info.getIntelDecoder16());
		assertEquals(tc.intelDecoder32, info.getIntelDecoder32());
		assertEquals(tc.intelDecoder64, info.getIntelDecoder64());
		assertEquals(tc.amdDecoder16, info.getAmdDecoder16());
		assertEquals(tc.amdDecoder32, info.getAmdDecoder32());
		assertEquals(tc.amdDecoder64, info.getAmdDecoder64());
		assertEquals(tc.table, info.getTable());
		assertEquals(tc.mandatoryPrefix, info.getMandatoryPrefix());
		assertEquals(tc.opCode, info.getOpCode());
		assertEquals(tc.opCodeLength, info.getOpCodeLength());
		assertEquals(tc.isGroup, info.isGroup());
		assertEquals(tc.groupIndex, info.getGroupIndex());
		assertEquals(tc.isRmGroup, info.isRmGroup());
		assertEquals(tc.rmGroupIndex, info.getRmGroupIndex());
		assertEquals(tc.opCount, info.getOpCount());
		assertEquals(tc.op0Kind, info.getOp0Kind());
		assertEquals(tc.op1Kind, info.getOp1Kind());
		assertEquals(tc.op2Kind, info.getOp2Kind());
		assertEquals(tc.op3Kind, info.getOp3Kind());
		assertEquals(tc.op4Kind, info.getOp4Kind());
		assertEquals(tc.op0Kind, info.getOpKind(0));
		assertEquals(tc.op1Kind, info.getOpKind(1));
		assertEquals(tc.op2Kind, info.getOpKind(2));
		assertEquals(tc.op3Kind, info.getOpKind(3));
		assertEquals(tc.op4Kind, info.getOpKind(4));
		assertEquals(5, IcedConstants.MAX_OP_COUNT);
		for (int i = tc.opCount; i < IcedConstants.MAX_OP_COUNT; i++)
			assertEquals(OpCodeOperandKind.NONE, info.getOpKind(i));
		assertEquals(tc.mvexEHBit, info.getMvexEHBit());
		assertEquals(tc.mvexCanUseEvictionHint, info.getMvexCanUseEvictionHint());
		assertEquals(tc.mvexCanUseImmRoundingControl, info.getMvexCanUseImmRoundingControl());
		assertEquals(tc.mvexIgnoresOpMaskRegister, info.getMvexIgnoresOpMaskRegister());
		assertEquals(tc.mvexNoSaeRc, info.getMvexNoSaeRc());
		assertEquals(tc.mvexTupleTypeLutKind, info.getMvexTupleTypeLutKind());
		assertEquals(tc.mvexConversionFunc, info.getMvexConversionFunc());
		assertEquals(tc.mvexValidConversionFuncsMask, info.getMvexValidConversionFuncsMask());
		assertEquals(tc.mvexValidSwizzleFuncsMask, info.getMvexValidSwizzleFuncsMask());
	}

	public static Iterable<Arguments> test_all_OpCodeInfos_Data() {
		OpCodeInfoTestCase[] allTests = OpCodeInfoTestCases.opCodeInfoTests;
		ArrayList<Arguments> result = new ArrayList<Arguments>(allTests.length);
		for (OpCodeInfoTestCase tc : allTests)
			result.add(Arguments.of(tc.lineNumber, tc.code, tc.opCodeString, tc.instructionString, tc));
		return result;
	}

	@Test
	void getOpKindThrowsIfInvalidInput() {
		OpCodeInfo info = Code.toOpCode(Code.AAA);
		assertThrows(IllegalArgumentException.class, () -> info.getOpKind(-0x8000_0000));
		assertThrows(IllegalArgumentException.class, () -> info.getOpKind(-1));
		info.getOpKind(0);
		info.getOpKind(IcedConstants.MAX_OP_COUNT - 1);
		assertThrows(IllegalArgumentException.class, () -> info.getOpKind(IcedConstants.MAX_OP_COUNT));
		assertThrows(IllegalArgumentException.class, () -> info.getOpKind(0x7FFF_FFFF));
	}

	@Test
	void verify_Instruction_OpCodeInfo() {
		for (int i = 0; i < IcedConstants.CODE_ENUM_COUNT; i++) {
			Instruction instruction = new Instruction();
			instruction.setCode(i);
			assertTrue(Code.toOpCode(i) == instruction.getOpCode());
		}
	}

	@Test
	void make_sure_all_Code_values_are_tested_exactly_once() {
		boolean[] tested = new boolean[IcedConstants.CODE_ENUM_COUNT];
		for (OpCodeInfoTestCase info : OpCodeInfoTestCases.opCodeInfoTests) {
			assertFalse(tested[info.code]);
			tested[info.code] = true;
		}
		StringBuilder sb = new StringBuilder();
		String[] codeNames = ToCode.names();
		for (int i = 0; i < tested.length; i++) {
			if (!tested[i] && !CodeUtils.isIgnored(codeNames[i])) {
				if (sb.length() > 0)
					sb.append(',');
				sb.append(codeNames[i]);
			}
		}
		assertEquals("", sb.toString());
	}
}
